// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/operation/CallbacksVisitor.mjs
import stampit from 'stampit';
import { isReferenceLikeElement } from "../../../predicates.mjs";
import { isReferenceElement } from "../../../../predicates.mjs";
import OperationCallbacksElement from "../../../../elements/nces/OperationCallbacks.mjs";
import MapVisitor from "../../generics/MapVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
const CallbacksVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Callback']
  },
  init() {
    this.element = new OperationCallbacksElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'callback');
      });
      return result;
    }
  }
});
export default CallbacksVisitor;
