// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/components/OperationTraitsVisitor.mjs
import stampit from 'stampit';
import ComponentsOperationTraitsElement from "../../../../elements/nces/ComponentsOperationTraits.mjs";
import MapVisitor from "../../generics/MapVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import { isReferenceLikeElement } from "../../../predicates.mjs";
import { isReferenceElement } from "../../../../predicates.mjs";
const OperationTraitsVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'OperationTrait'];
    }
  },
  init() {
    this.element = new ComponentsOperationTraitsElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'operationTrait');
      });
      return result;
    }
  }
});
export default OperationTraitsVisitor;
