// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/parameter/index.mjs
import stampit from 'stampit';
import { always } from 'ramda';
import { isObjectElement, toValue } from '@swagger-api/apidom-core';
import ParameterElement from "../../../../elements/Parameter.mjs";
import FixedFieldsVisitor from "../../generics/FixedFieldsVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import { isMediaTypeElement } from "../../../../predicates.mjs";
const ParameterVisitor = stampit(FixedFieldsVisitor, FallbackVisitor, {
  props: {
    specPath: always(['document', 'objects', 'Parameter']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new ParameterElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every MediaTypeElement with media type metadata
      if (isObjectElement(this.element.contentProp)) {
        this.element.contentProp.filter(isMediaTypeElement).forEach((mediaTypeElement, key) => {
          mediaTypeElement.setMetaProperty('media-type', toValue(key));
        });
      }
      return result;
    }
  }
});
export default ParameterVisitor;
