// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/components/ExamplesVisitor.mjs
import stampit from 'stampit';
import ComponentsExamplesElement from "../../../../elements/nces/ComponentsExamples.mjs";
import MapVisitor from "../../generics/MapVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import { isReferenceLikeElement } from "../../../predicates.mjs";
import { isReferenceElement } from "../../../../predicates.mjs";
const ExamplesVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Example'];
    }
  },
  init() {
    this.element = new ComponentsExamplesElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every ReferenceElement with metadata about their referencing type
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'example');
      });
      return result;
    }
  }
});
export default ExamplesVisitor;
