// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/components/LinksVisitor.mjs
import stampit from 'stampit';
import ComponentsLinksElement from "../../../../elements/nces/ComponentsLinks.mjs";
import MapVisitor from "../../generics/MapVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import { isReferenceLikeElement } from "../../../predicates.mjs";
import { isReferenceElement } from "../../../../predicates.mjs";
const LinksVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Link'];
    }
  },
  init() {
    this.element = new ComponentsLinksElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'link');
      });
      return result;
    }
  }
});
export default LinksVisitor;
