// @swagger-api/apidom-ns-openapi-2/cjs/refractor/visitors/open-api-2/definitions/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _apidomNsJsonSchemaDraft = require("@swagger-api/apidom-ns-json-schema-draft-4");
var _Definitions = _interopRequireDefault(require("../../../../elements/Definitions.cjs"));
var _MapVisitor = _interopRequireDefault(require("../../generics/MapVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
const DefinitionsVisitor = (0, _stampit.default)(_MapVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: element => {
      return (0, _apidomNsJsonSchemaDraft.isJSONReferenceLikeElement)(element) ? ['document', 'objects', 'JSONReference'] : ['document', 'objects', 'Schema'];
    }
  },
  init() {
    this.element = new _Definitions.default();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _MapVisitor.default.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every JSONReferenceElement with metadata about their referencing type
      this.element.filter(_apidomNsJsonSchemaDraft.isJSONReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'schema');
      });
      return result;
    }
  }
});
var _default = exports.default = DefinitionsVisitor;
