// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/encoding/index.mjs
import stampit from 'stampit';
import { always } from 'ramda';
import { isObjectElement, toValue } from '@swagger-api/apidom-core';
import EncodingElement from "../../../../elements/Encoding.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import FixedFieldsVisitor from "../../generics/FixedFieldsVisitor.mjs";
import { isHeaderElement } from "../../../../predicates.mjs";
const EncodingVisitor = stampit(FixedFieldsVisitor, FallbackVisitor, {
  props: {
    specPath: always(['document', 'objects', 'Encoding']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new EncodingElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every Header with media type metadata
      if (isObjectElement(this.element.headers)) {
        this.element.headers.filter(isHeaderElement).forEach((headerElement, key) => {
          headerElement.setMetaProperty('header-name', toValue(key));
        });
      }
      return result;
    }
  }
});
export default EncodingVisitor;
