// @swagger-api/apidom-ns-openapi-2/cjs/refractor/visitors/open-api-2/paths/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _apidomCore = require("@swagger-api/apidom-core");
var _Paths = _interopRequireDefault(require("../../../../elements/Paths.cjs"));
var _PatternedFieldsVisitor = _interopRequireDefault(require("../../generics/PatternedFieldsVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _predicates = require("../../../../predicates.cjs");
const PathsVisitor = (0, _stampit.default)(_PatternedFieldsVisitor.default, _FallbackVisitor.default, {
  props: {
    fieldPatternPredicate: _ramda.T,
    specPath: (0, _ramda.always)(['document', 'objects', 'PathItem']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new _Paths.default();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _PatternedFieldsVisitor.default.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every PathItemElement with path metadata
      this.element.filter(_predicates.isPathItemElement).forEach((pathItemElement, key) => {
        key.classes.push('openapi-path-template');
        key.classes.push('path-template');
        pathItemElement.setMetaProperty('path', (0, _apidomCore.cloneDeep)(key));
      });
      return result;
    }
  }
});
var _default = exports.default = PathsVisitor;
