// @swagger-api/apidom-ns-openapi-3-0/cjs/refractor/visitors/open-api-3-0/components/ResponsesVisitor.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _apidomCore = require("@swagger-api/apidom-core");
var _ComponentsResponses = _interopRequireDefault(require("../../../../elements/nces/ComponentsResponses.cjs"));
var _MapVisitor = _interopRequireDefault(require("../../generics/MapVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _predicates = require("../../../predicates.cjs");
var _predicates2 = require("../../../../predicates.cjs");
const ResponsesVisitor = (0, _stampit.default)(_MapVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: element => {
      return (0, _predicates.isReferenceLikeElement)(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Response'];
    }
  },
  init() {
    this.element = new _ComponentsResponses.default();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _MapVisitor.default.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every ReferenceElement with metadata about their referencing type
      this.element.filter(_predicates2.isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'response');
      });

      // decorate every ResponseElement with metadata about their status code
      this.element.filter(_predicates2.isResponseElement).forEach((value, key) => {
        value.setMetaProperty('http-status-code', (0, _apidomCore.toValue)(key));
      });
      return result;
    }
  }
});
var _default = exports.default = ResponsesVisitor;
