// @swagger-api/apidom-ns-openapi-2/es/refractor/visitors/open-api-2/path-item/index.mjs
import stampit from 'stampit';
import { always } from 'ramda';
import { isStringElement, cloneDeep, toValue } from '@swagger-api/apidom-core';
import PathItemElement from "../../../../elements/PathItem.mjs";
import { isOperationElement } from "../../../../predicates.mjs";
import FixedFieldsVisitor from "../../generics/FixedFieldsVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
const PathItemVisitor = stampit(FixedFieldsVisitor, FallbackVisitor, {
  props: {
    specPath: always(['document', 'objects', 'PathItem'])
  },
  init() {
    this.element = new PathItemElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // decorate Operation elements with HTTP method
      this.element.filter(isOperationElement).forEach((operationElement, httpMethodElementCI) => {
        const httpMethodElementCS = cloneDeep(httpMethodElementCI);
        httpMethodElementCS.content = toValue(httpMethodElementCS).toUpperCase();
        operationElement.setMetaProperty('http-method', httpMethodElementCS);
      });

      // mark this PathItemElement with reference metadata
      if (isStringElement(this.element.$ref)) {
        this.element.classes.push('reference-element');
      }
      return result;
    }
  }
});
export default PathItemVisitor;
