// @swagger-api/apidom-ast/cjs/json/nodes/predicates.cjs
"use strict";

exports.__esModule = true;
exports.isTrue = exports.isStringContent = exports.isString = exports.isProperty = exports.isObject = exports.isNumber = exports.isNull = exports.isKey = exports.isFalse = exports.isEscapeSequence = exports.isDocument = exports.isArray = void 0;
var _predicates = require("../../predicates.cjs");
const isDocument = exports.isDocument = _predicates.isNodeType.bind(undefined, 'document');
const isString = exports.isString = _predicates.isNodeType.bind(undefined, 'string');
const isFalse = exports.isFalse = _predicates.isNodeType.bind(undefined, 'false');
const isTrue = exports.isTrue = _predicates.isNodeType.bind(undefined, 'true');
const isNull = exports.isNull = _predicates.isNodeType.bind(undefined, 'null');
const isNumber = exports.isNumber = _predicates.isNodeType.bind(undefined, 'number');
const isArray = exports.isArray = _predicates.isNodeType.bind(undefined, 'array');
const isObject = exports.isObject = _predicates.isNodeType.bind(undefined, 'object');
const isStringContent = exports.isStringContent = _predicates.isNodeType.bind(undefined, 'stringContent');
const isEscapeSequence = exports.isEscapeSequence = _predicates.isNodeType.bind(undefined, 'escapeSequence');
const isProperty = exports.isProperty = _predicates.isNodeType.bind(undefined, 'property');
const isKey = exports.isKey = _predicates.isNodeType.bind(undefined, 'key');
