// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/paths/index.mjs
import stampit from 'stampit';
import { T as stubTrue, always } from 'ramda';
import { cloneDeep } from '@swagger-api/apidom-core';
import PathsElement from "../../../../elements/Paths.mjs";
import PatternedFieldsVisitor from "../../generics/PatternedFieldsVisitor.mjs";
import FallbackVisitor from "../../FallbackVisitor.mjs";
import { isPathItemElement } from "../../../../predicates.mjs";
const PathsVisitor = stampit(PatternedFieldsVisitor, FallbackVisitor, {
  props: {
    fieldPatternPredicate: stubTrue,
    specPath: always(['document', 'objects', 'PathItem']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new PathsElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = PatternedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every PathItemElement with path metadata
      this.element.filter(isPathItemElement).forEach((pathItemElement, key) => {
        key.classes.push('openapi-path-template');
        key.classes.push('path-template');
        pathItemElement.setMetaProperty('path', cloneDeep(key));
      });
      return result;
    }
  }
});
export default PathsVisitor;
