// @swagger-api/apidom-ast/cjs/yaml/nodes/predicates.cjs
"use strict";

exports.__esModule = true;
exports.isTag = exports.isStream = exports.isSequence = exports.isScalar = exports.isMapping = exports.isKeyValuePair = exports.isDocument = exports.isDirective = exports.isComment = exports.isAlias = void 0;
var _predicates = require("../../predicates.cjs");
const isStream = exports.isStream = _predicates.isNodeType.bind(undefined, 'stream');
const isDocument = exports.isDocument = _predicates.isNodeType.bind(undefined, 'document');
const isMapping = exports.isMapping = _predicates.isNodeType.bind(undefined, 'mapping');
const isSequence = exports.isSequence = _predicates.isNodeType.bind(undefined, 'sequence');
const isKeyValuePair = exports.isKeyValuePair = _predicates.isNodeType.bind(undefined, 'keyValuePair');
const isTag = exports.isTag = _predicates.isNodeType.bind(undefined, 'tag');
const isScalar = exports.isScalar = _predicates.isNodeType.bind(undefined, 'scalar');
const isAlias = exports.isAlias = _predicates.isNodeType.bind(undefined, 'alias');
const isDirective = exports.isDirective = _predicates.isNodeType.bind(undefined, 'directive');
const isComment = exports.isComment = _predicates.isNodeType.bind(undefined, 'comment');
